File: src/main/java/fr/vlegall/sochief/model/recipe/RecipeStep.java
package fr.vlegall.sochief.model.recipe;

import com.fasterxml.jackson.annotation.JsonIgnore;
import fr.vlegall.sochief.model.common.AuditableEntity;
import jakarta.persistence.*;
import jakarta.validation.constraints.AssertTrue;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Duration;

@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "RECIPE_STEP")
public class RecipeStep extends AuditableEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull(message = "Recipe cannot be null.")
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "recipe_id", nullable = false)
    @JsonIgnore
    private Recipe recipe;

    @Column(columnDefinition = "TEXT")
    private String description;

    @NotNull(message = "Duration cannot be null.")
    @Column(nullable = false, name = "duration")
    private Duration duration;

    @NotNull(message = "Position cannot be null.")
    @Positive
    @Column(nullable = false, name = "position")
    private Integer position;

    @AssertTrue(message = "Duration must be positive.")
    public boolean isDurationPositive() {
        return duration != null && duration.compareTo(Duration.ZERO) > 0;
    }
}

File: src/main/kotlin/com/vlegall/sochief/model/recipe/RecipeStep.kt
package com.vlegall.sochief.model.recipe

import com.fasterxml.jackson.annotation.JsonIgnore
import com.vlegall.sochief.model.common.AuditableEntity
import jakarta.persistence.*
import jakarta.validation.constraints.AssertTrue
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Positive
import java.time.Duration

@Entity
@Table(name = "RECIPE_STEP")
class RecipeStep(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @field:NotNull(message = "Recipe cannot be null.")
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "recipe_id", nullable = false)
    @JsonIgnore
    var recipe: Recipe? = null,

    @Column(columnDefinition = "TEXT")
    var description: String? = null,

    @field:NotNull(message = "Duration cannot be null.")
    @Column(nullable = false, name = "duration")
    var duration: Duration? = null,

    @field:NotNull(message = "Position cannot be null.")
    @field:Positive
    @Column(nullable = false, name = "position")
    var position: Int? = null,
) : AuditableEntity() {
    @AssertTrue(message = "Duration must be positive.")
    fun isDurationPositive(): Boolean = duration!! > Duration.ZERO
}